#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
data_path = '../../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')['gross income'].sum()

Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64

Сумарний прибуток найбільший у філіала С = 5265.1765

2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [11]:
df.groupby('Branch')['City'].unique()

Branch
A       [Yangon]
B     [Mandalay]
C    [Naypyitaw]
Name: City, dtype: object

Філіал С з найвищим прибутком знаходиться в місті Naypyitaw

3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [16]:
invoices_by_category = df.pivot_table(index = 'Branch', columns = 'Product line', values = 'Invoice ID', 
                        aggfunc = 'size')

In [18]:
invoices_by_category

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [44]:
invoices_by_category.groupby('Branch')['Electronic accessories'].max()

Branch
A    60
B    55
C    55
Name: Electronic accessories, dtype: int64

4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


4. Хочу проаналізувати по містах хто більше купує - чоловіки чи жінки. Потім проаналізувати які категорії товарів найбільш популярні серед чоловіків та жінок, і відштовхуючись від цієї інформації надати рекомендації по розширенню асортименту та збільшення частки певних категорій товарів.

In [25]:
df.pivot_table(index = 'Gender', columns = 'City', values = 'Invoice ID', aggfunc = 'size') 
# подивимось по кожному місту кількість покупок чоловіками та жінками

City    Mandalay  Naypyitaw  Yangon
Gender                             
Female       162        178     161
Male         170        150     179

In [28]:
gender_product = df.pivot_table(index = 'Gender', columns = 'Product line', values = 'Invoice ID', 
                        aggfunc = 'size') 
# подивимось на розподіл кількості закупів чоловіками та жінками між категоріями товарів

In [29]:
gender_product

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Gender                                                                          
Female                            84                   96                  90   
Male                              86                   82                  84   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Gender                                                                  
Female                       64                  79                 88  
Male                         88                  81                 78

In [31]:
gender_product.loc['Female'].idxmax() #знаходжу яку категорію обирає більшість жінок

'Fashion accessories'

In [37]:
gender_product.drop(columns = 'Fashion accessories').loc['Female'].idxmax() #знаходжу 2гу за популярністю категорію

'Food and beverages'

    Так само знаходжу найбільш попурні групи товарів серед чоловіків:

In [38]:
gender_product.loc['Male'].idxmax()

'Health and beauty'

In [40]:
gender_product.drop(columns = 'Health and beauty').loc['Male'].idxmax()

'Electronic accessories'

    В містах Mandalay та Yangon більшість покупок здійснюють чоловіки. Проаналізувавши вподобання чоловіків, бачимо, що найбільшою популярністю користуються групи товарів 'Health and beauty' та 'Electronic accessories'. Отже можемо надати рекомендацію для супермаркетів цих міст зробити акцент на розширення асортименту товарів 'Health and beauty' та 'Electronic accessories. Також ввести лінійку більш дорогих товарів із цих категорій.
    В місті Naypyitaw основні клієнти - жінки. Згідно отриманої інформації найбільшим попитом серед цього сегменту клієнтів користуються товари групи 'Fashion accessories' та 'Food and beverages'. Тобто потрібно розширити асортимент та більш дорогі товари саме цих груп товарів.

5. Проаналізуємо по місту Yangon, які групи товарів приносять максимум прибутку і як цей прибуток розподіляється усередині кожної групи

In [57]:
df[df.City == 'Yangon'].groupby('Product line')['Total'].describe()

count        mean         std      min         25%  \
Product line                                                                 
Electronic accessories   60.0  305.285225  246.411455  30.4080  101.490375   
Fashion accessories      51.0  320.245265  257.255822  12.6945   94.179750   
Food and beverages       58.0  295.915526  205.761806  33.4320  165.193875   
Health and beauty        47.0  268.037298  186.079373  19.2465  135.549750   
Home and lifestyle       65.0  344.879931  246.511642  19.1940  147.672000   
Sports and travel        59.0  328.350839  235.076176  16.1070  148.527750   

                             50%         75%        max  
Product line                                             
Electronic accessories  207.0705  442.622250   931.0350  
Fashion accessories     277.6725  485.173500  1039.2900  
Food and beverages      249.0705  356.346375   932.3370  
Health and beauty       217.1820  370.980750   752.6400  
Home and lifestyle      263.1300  493.794000   951.8250  
Sports and travel       271.2780  465.990000   926.9505

    Виходячи з отриманих даних, бачимо, що найбільший дохід приносить група Fashion accessories та Home and lifestyle, отже саме ці напрямки можна активно розвивати. Найменший дохід приносить група товарів Health and beauty - можливо потрібно провести додаткові маркетингові заходи для привернення уваги до цих товарів.
    Також цікаво, що різниця між мах та 75 персинталем - більше половини від загальної виручки. Можна припустити, що населення міста Yangon - заможне і обирає більш дорогі товари. Слід розширити асортимент саме з доварами з високою ціною.

6. Хочу дослідити поведінку різних Customer type: чи середній чек Member буде більшим, ніж чек Normal, наприклад по місту Mandalay.

In [58]:
df[df.City == 'Mandalay'].groupby('Customer type')['Total'].agg(['mean', 'max', 'min'])

mean       max      min
Customer type                               
Member         325.482945   944.622  20.1075
Normal         314.329257  1022.490  18.6375

    Бачимо, що середній чек Member трохи більший (~3.43 %), ніж чек Normal, а максимальний - навіть менший. Отже потрібно проводити більш активні маркетингові заходи серед Member.